daemon/src/cfd.rs

```diff
@@ -0,0 +1,113 @@
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Usd(pub u64);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Leverage(u8);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum TradingPair {
+    BtcUsd,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum Position {
+    Buy,
+    Sell,
+}
+
+/// A concrete offer for a user
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CfdOffer {
+    pub id: Uuid,
+
+    pub price: Usd,
+
+    // TODO: [post-MVP] Representation of the contract size; at the moment the contract size is always 1 USD
+    pub min_amount: Usd,
+    pub max_amount: Usd,
+
+    // TODO: [post-MVP] Allow different values
+    pub leverage: Leverage,
+    pub trading_pair: TradingPair,
+
+    // TODO: [post-MVP] This is dynamic based on the price and leverage (might want to created packages for this or calculate on UI side)
+    pub liquidation_price: Usd,
+}
+
+/// The taker POSTs this to create a Cfd
+#[derive(Debug, Clone, Deserialize)]
+pub struct CfdTakeRequest {
+    pub offer_id: Uuid,
+    pub quantity: Usd,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum Error {
+    // TODO
+    ConnectionLost,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CfdStateError {
+    last_successful_state: CfdState,
+    error: Error,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum CfdState {
+    /// The taker has sent an open request
+    TakeRequested,
+    /// We have sent a request to the maker to open the CFD but don't have a response yet
+    PendingOpen,
+    /// The maker has agreed to the CFD, but the contract is not set up on chain yet
+    Accepted,
+    /// The CFD contract is set up on chain
+    Open,
+    /// The taker has requested to close the position, but we have not passed that on to the blockchain yet.
+    CloseRequested,
+    /// The close transaction (CET) was published on the Bitcoin blockchain but we don't have a confirmation yet.
+    PendingClose,
+    /// The close transaction is confirmed with at least one block
+    Closed,
+    // FIXME: add CfdError
+    Error,
+}
+
+/// Represents a cfd (including state)
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Cfd {
+    pub initial_price: Usd,
+
+    pub leverage: Leverage,
+    pub trading_pair: TradingPair,
+    pub liquidation_price: Usd,
+
+    #[serde(with = "::bitcoin::util::amount::serde::as_btc")]
+    pub quantity_btc: Amount,
+    pub quantity_usd: Usd,
+
+    #[serde(with = "::bitcoin::util::amount::serde::as_btc")]
+    pub profit_btc: Amount,
+    pub profit_usd: Usd,
+
+    pub creation_date: SystemTime,
+
+    pub state: CfdState,
+}
+
+pub fn static_cfd_offer() -> CfdOffer {
+    CfdOffer {
+        id: Uuid::new_v4(),
+        price: Usd(49_000),
+        min_amount: Usd(1_000),
+        max_amount: Usd(10_000),
+        leverage: Leverage(5),
+        trading_pair: TradingPair::BtcUsd,
+        liquidation_price: Usd(42_000),
```

Comment:
Just so that this is not hit you later on: the `liquidation_price` depends on the taker's input (amount + leverage), e.g. if he wants to buy 1 BTC at $49k with a leverage of 5 the needed collateral will be 0.2BTC and liquidation_price depends on this amount.

Reply:
As far as our formulas in the model go, the liquidation price is purely based on the `leverage` factor (i.e. `x5`) and the initial price (and a maintenance margin that we currently don't depict yet). 
The amount the taker enters is irrelevant as far as I understand our model!

Given, that we have a fixed leverage at the moment the liquidation price is fixed for each offer, because the offer includes price and leverage.

Reply:
The taker can enter an amount between `min_amount` and `max_amount` right? If you look at our spread sheet: that's the quantity the taker wants to buy. That's what the maker (as the lender or the person who wants to go short) then has to come up with. 
This amount + the leverage (`5x`) influences the `liquidation_price`. 

Differently said: 
- the maker has to put up an amount X in BTC which the taker wants to buy (defined in USD aka Z). 
- the taker has to put up an amount Y in BTC which depends on the `X / leverage`
- the `liquidation_price` depends then on the combination of `X` and `Y`

I hope that helps.

Reply:
Don't mind me then :D As discussed via call.

Reply:
Note for completion: the model is currently based the formula described as `isolated margin mode` described [here](
https://help.bybit.com/hc/en-us/articles/360039261334-How-to-calculate-Liquidation-Price-Inverse-Contract-).